File: windows/iCam.Core/Protocol/CameraState.cs
using System.Text.Json.Serialization;

namespace ICam.Core.Protocol;

[JsonConverter(typeof(JsonStringEnumConverter<VideoCodec>))]
public enum VideoCodec { [JsonStringEnumMemberName("h264")] H264, [JsonStringEnumMemberName("hevc")] Hevc }

[JsonConverter(typeof(JsonStringEnumConverter<HdrMode>))]
public enum HdrMode { [JsonStringEnumMemberName("off")] Off, [JsonStringEnumMemberName("auto")] Auto, [JsonStringEnumMemberName("on")] On }

[JsonConverter(typeof(JsonStringEnumConverter<ExposureMode>))]
public enum ExposureMode { [JsonStringEnumMemberName("auto")] Auto, [JsonStringEnumMemberName("manual")] Manual }

[JsonConverter(typeof(JsonStringEnumConverter<WhiteBalanceMode>))]
public enum WhiteBalanceMode { [JsonStringEnumMemberName("auto")] Auto, [JsonStringEnumMemberName("manual")] Manual }

[JsonConverter(typeof(JsonStringEnumConverter<FocusMode>))]
public enum FocusMode
{
    [JsonStringEnumMemberName("continuous")] Continuous,
    [JsonStringEnumMemberName("single")] Single,
    [JsonStringEnumMemberName("manual")] Manual,
}

[JsonConverter(typeof(JsonStringEnumConverter<Stabilization>))]
public enum Stabilization
{
    [JsonStringEnumMemberName("off")] Off,
    [JsonStringEnumMemberName("standard")] Standard,
    [JsonStringEnumMemberName("cinematic")] Cinematic,
    [JsonStringEnumMemberName("cinematicExtended")] CinematicExtended,
    [JsonStringEnumMemberName("auto")] Auto,
}

[JsonConverter(typeof(JsonStringEnumConverter<TorchMode>))]
public enum TorchMode { [JsonStringEnumMemberName("off")] Off, [JsonStringEnumMemberName("on")] On, [JsonStringEnumMemberName("auto")] Auto }

[JsonConverter(typeof(JsonStringEnumConverter<OrientationLock>))]
public enum OrientationLock
{
    [JsonStringEnumMemberName("auto")] Auto,
    [JsonStringEnumMemberName("portrait")] Portrait,
    [JsonStringEnumMemberName("landscapeLeft")] LandscapeLeft,
    [JsonStringEnumMemberName("landscapeRight")] LandscapeRight,
}

[JsonConverter(typeof(JsonStringEnumConverter<CaptureTarget>))]
public enum CaptureTarget
{
    [JsonStringEnumMemberName("phone")] Phone,
    [JsonStringEnumMemberName("pc")] Pc,
    [JsonStringEnumMemberName("both")] Both,
}

[JsonConverter(typeof(JsonStringEnumConverter<WhiteBalancePreset>))]
public enum WhiteBalancePreset
{
    [JsonStringEnumMemberName("auto")] Auto,
    [JsonStringEnumMemberName("daylight")] Daylight,
    [JsonStringEnumMemberName("cloudy")] Cloudy,
    [JsonStringEnumMemberName("shade")] Shade,
    [JsonStringEnumMemberName("tungsten")] Tungsten,
    [JsonStringEnumMemberName("fluorescent")] Fluorescent,
    [JsonStringEnumMemberName("warm")] Warm,
    [JsonStringEnumMemberName("custom")] Custom,
}

/// <summary>
/// The camera state, as this computer last received it.
///
/// Windows renders this and sends mutations. It never treats its own copy as
/// authoritative — the phone owns the state, because only the phone can see
/// what the hardware actually accepted.
/// </summary>
public sealed record CameraState
{
    [JsonPropertyName("v")] public ulong Version { get; init; }

    [JsonPropertyName("lensId")] public string LensId { get; init; } = "";
    [JsonPropertyName("zoom")] public double Zoom { get; init; } = 1.0;
    [JsonPropertyName("lensLocked")] public bool LensLocked { get; init; }

    [JsonPropertyName("width")] public int Width { get; init; } = 1920;
    [JsonPropertyName("height")] public int Height { get; init; } = 1080;
    [JsonPropertyName("fps")] public int Fps { get; init; } = 30;
    [JsonPropertyName("codec")] public VideoCodec Codec { get; init; } = VideoCodec.Hevc;
    [JsonPropertyName("hdr")] public HdrMode Hdr { get; init; } = HdrMode.Auto;
    [JsonPropertyName("stabilization")] public Stabilization Stabilization { get; init; } = Stabilization.Auto;

    [JsonPropertyName("exposureMode")] public ExposureMode ExposureMode { get; init; } = ExposureMode.Auto;
    [JsonPropertyName("iso")] public double Iso { get; init; } = 100;
    [JsonPropertyName("exposureDurationUs")] public int ExposureDurationUs { get; init; } = 8333;
    [JsonPropertyName("ev")] public double Ev { get; init; }
    [JsonPropertyName("exposureLocked")] public bool ExposureLocked { get; init; }

    [JsonPropertyName("whiteBalanceMode")] public WhiteBalanceMode WhiteBalanceMode { get; init; } = WhiteBalanceMode.Auto;
    [JsonPropertyName("whiteBalancePreset")] public WhiteBalancePreset WhiteBalancePreset { get; init; } = WhiteBalancePreset.Auto;
    [JsonPropertyName("temperature")] public double Temperature { get; init; } = 5000;
    [JsonPropertyName("tint")] public double Tint { get; init; }

    [JsonPropertyName("focusMode")] public FocusMode FocusMode { get; init; } = FocusMode.Continuous;
    [JsonPropertyName("focusPosition")] public double FocusPosition { get; init; } = 0.5;
    [JsonPropertyName("focusLocked")] public bool FocusLocked { get; init; }
    [JsonPropertyName("faceDrivenFocus")] public bool FaceDrivenFocus { get; init; } = true;

    [JsonPropertyName("torch")] public TorchMode Torch { get; init; } = TorchMode.Off;
    [JsonPropertyName("torchLevel")] public double TorchLevel { get; init; } = 1.0;
    [JsonPropertyName("mirrored")] public bool Mirrored { get; init; }
    [JsonPropertyName("orientation")] public OrientationLock Orientation { get; init; } = OrientationLock.Auto;

    // Applied to the derived outputs on the PC, never to the master recording.
    // Each runs -1 to +1 and means "leave it alone" at 0, except lowLight and
    // beauty, which only have one direction to go and run 0 to 1.
    [JsonPropertyName("brightness")] public double Brightness { get; init; }
    [JsonPropertyName("contrast")] public double Contrast { get; init; }
    [JsonPropertyName("saturation")] public double Saturation { get; init; }
    [JsonPropertyName("warmth")] public double Warmth { get; init; }
    [JsonPropertyName("sharpness")] public double Sharpness { get; init; }
    [JsonPropertyName("lowLight")] public double LowLight { get; init; }
    [JsonPropertyName("beauty")] public double Beauty { get; init; }

    /// <summary>8333 µs reads as 1/120.</summary>
    [JsonIgnore]
    public int ShutterDenominator =>
        ExposureDurationUs > 0 ? (int)Math.Round(1_000_000.0 / ExposureDurationUs) : 0;
}

/// <summary>
/// A partial change. Only the properties the sender actually touched are
/// serialised, which is what makes concurrent edits from the phone and the PC
/// safe: a stale ISO slider cannot revert an unrelated setting.
/// </summary>
public sealed class CameraMutation
{
    [JsonPropertyName("lensId")] public string? LensId { get; set; }
    [JsonPropertyName("zoom")] public double? Zoom { get; set; }
    [JsonPropertyName("lensLocked")] public bool? LensLocked { get; set; }

    [JsonPropertyName("width")] public int? Width { get; set; }
    [JsonPropertyName("height")] public int? Height { get; set; }
    [JsonPropertyName("fps")] public int? Fps { get; set; }
    [JsonPropertyName("codec")] public VideoCodec? Codec { get; set; }
    [JsonPropertyName("hdr")] public HdrMode? Hdr { get; set; }
    [JsonPropertyName("stabilization")] public Stabilization? Stabilization { get; set; }

    [JsonPropertyName("exposureMode")] public ExposureMode? ExposureMode { get; set; }
    [JsonPropertyName("iso")] public double? Iso { get; set; }
    [JsonPropertyName("exposureDurationUs")] public int? ExposureDurationUs { get; set; }
    [JsonPropertyName("ev")] public double? Ev { get; set; }
    [JsonPropertyName("exposureLocked")] public bool? ExposureLocked { get; set; }

    [JsonPropertyName("whiteBalanceMode")] public WhiteBalanceMode? WhiteBalanceMode { get; set; }
    [JsonPropertyName("whiteBalancePreset")] public WhiteBalancePreset? WhiteBalancePreset { get; set; }
    [JsonPropertyName("temperature")] public double? Temperature { get; set; }
    [JsonPropertyName("tint")] public double? Tint { get; set; }

    [JsonPropertyName("focusMode")] public FocusMode? FocusMode { get; set; }
    [JsonPropertyName("focusPosition")] public double? FocusPosition { get; set; }
    [JsonPropertyName("focusLocked")] public bool? FocusLocked { get; set; }
    [JsonPropertyName("faceDrivenFocus")] public bool? FaceDrivenFocus { get; set; }

    [JsonPropertyName("torch")] public TorchMode? Torch { get; set; }
    [JsonPropertyName("torchLevel")] public double? TorchLevel { get; set; }
    [JsonPropertyName("mirrored")] public bool? Mirrored { get; set; }
    [JsonPropertyName("orientation")] public OrientationLock? Orientation { get; set; }

    [JsonPropertyName("brightness")] public double? Brightness { get; set; }
    [JsonPropertyName("contrast")] public double? Contrast { get; set; }
    [JsonPropertyName("saturation")] public double? Saturation { get; set; }
    [JsonPropertyName("warmth")] public double? Warmth { get; set; }
    [JsonPropertyName("sharpness")] public double? Sharpness { get; set; }
    [JsonPropertyName("lowLight")] public double? LowLight { get; set; }
    [JsonPropertyName("beauty")] public double? Beauty { get; set; }
}

/// <summary>What this computer receives. Unrelated to what the phone records.</summary>
public sealed record StreamProfile
{
    [JsonPropertyName("width")] public int Width { get; init; } = 1280;
    [JsonPropertyName("height")] public int Height { get; init; } = 720;
    [JsonPropertyName("fps")] public int Fps { get; init; } = 30;
    [JsonPropertyName("codec")] public VideoCodec Codec { get; init; } = VideoCodec.H264;
    [JsonPropertyName("bitrate")] public int Bitrate { get; init; } = 6_000_000;
    [JsonPropertyName("keyframeIntervalSeconds")] public double KeyframeIntervalSeconds { get; init; } = 2;

    public static StreamProfile Webcam720p30 => new()
    { Width = 1280, Height = 720, Fps = 30, Codec = VideoCodec.H264, Bitrate = 4_000_000 };

    public static StreamProfile Webcam1080p30 => new()
    { Width = 1920, Height = 1080, Fps = 30, Codec = VideoCodec.H264, Bitrate = 8_000_000 };

    public static StreamProfile Webcam1080p60 => new()
    { Width = 1920, Height = 1080, Fps = 60, Codec = VideoCodec.H264, Bitrate = 12_000_000 };

    /// <summary>
    /// H.264 rather than HEVC, deliberately: HEVC decode on Windows depends on
    /// a store extension half the machines do not have, and a webcam that
    /// only works on some computers is not a webcam.
    /// </summary>
    public static StreamProfile Webcam2160p30 => new()
    { Width = 3840, Height = 2160, Fps = 30, Codec = VideoCodec.H264, Bitrate = 20_000_000 };
}

File: ios/iCam/Camera/CameraState.swift
import Foundation

// MARK: - Vocabulary shared with Windows

enum VideoCodec: String, Codable, Sendable, CaseIterable {
    case h264, hevc
    var displayName: String { self == .h264 ? "H.264" : "HEVC" }
}

enum HDRMode: String, Codable, Sendable, CaseIterable {
    case off, auto, on
}

enum ExposureMode: String, Codable, Sendable, CaseIterable {
    case auto, manual
}

enum WhiteBalanceMode: String, Codable, Sendable, CaseIterable {
    case auto, manual
}

enum FocusMode: String, Codable, Sendable, CaseIterable {
    case continuous, single, manual
}

enum Stabilization: String, Codable, Sendable, CaseIterable {
    case off, standard, cinematic, cinematicExtended, auto
}

enum TorchMode: String, Codable, Sendable, CaseIterable {
    case off, on, auto
}

enum OrientationLock: String, Codable, Sendable, CaseIterable {
    case auto, portrait, landscapeLeft, landscapeRight
}

enum CaptureTarget: String, Codable, Sendable, CaseIterable {
    case phone, pc, both
}

/// White balance presets, expressed as a temperature the device can actually
/// be driven to. `custom` means the user picked a value or used Pick White.
enum WhiteBalancePreset: String, Codable, Sendable, CaseIterable {
    case auto, daylight, cloudy, shade, tungsten, fluorescent, warm, custom

    /// Kelvin, or `nil` for the two presets that are not a fixed temperature.
    var temperature: Double? {
        switch self {
        case .auto, .custom: return nil
        case .daylight:      return 5600
        case .cloudy:        return 6500
        case .shade:         return 7500
        case .tungsten:      return 3200
        case .fluorescent:   return 4000
        case .warm:          return 2800
        }
    }
}

// MARK: - The authoritative state

/// The one camera state. It lives on the phone, inside `CameraStateStore`.
/// Windows renders the last copy it received and sends mutations; it never
/// keeps a second state it believes in.
struct CameraState: Codable, Equatable, Sendable {
    var version: UInt64 = 0

    // Optics
    var lensId: String = ""
    var zoom: Double = 1.0
    var lensLocked: Bool = false

    // Format
    var width: Int = 1920
    var height: Int = 1080
    var fps: Int = 30
    var codec: VideoCodec = .hevc
    var hdr: HDRMode = .auto
    var stabilization: Stabilization = .auto

    // Exposure
    var exposureMode: ExposureMode = .auto
    var iso: Double = 100
    var exposureDurationUs: Int = 8_333
    var ev: Double = 0
    var exposureLocked: Bool = false

    // Colour
    var whiteBalanceMode: WhiteBalanceMode = .auto
    var whiteBalancePreset: WhiteBalancePreset = .auto
    var temperature: Double = 5000
    var tint: Double = 0

    // Focus
    var focusMode: FocusMode = .continuous
    var focusPosition: Double = 0.5
    var focusLocked: Bool = false
    var faceDrivenFocus: Bool = true

    // Output
    var torch: TorchMode = .off
    var torchLevel: Double = 1.0
    var mirrored: Bool = false
    var orientation: OrientationLock = .auto

    // Image rendering (applied to derived outputs, never to the master).
    // Each runs -1 to +1 and means "leave it alone" at 0.
    var brightness: Double = 0
    var contrast: Double = 0
    var saturation: Double = 0
    var warmth: Double = 0
    var sharpness: Double = 0

    enum CodingKeys: String, CodingKey {
        case version = "v"
        case lensId, zoom, lensLocked
        case width, height, fps, codec, hdr, stabilization
        case exposureMode, iso, exposureDurationUs, ev, exposureLocked
        case whiteBalanceMode, whiteBalancePreset, temperature, tint
        case focusMode, focusPosition, focusLocked, faceDrivenFocus
        case torch, torchLevel, mirrored, orientation
        case brightness, contrast, saturation, warmth, sharpness
    }

    /// Shutter denominator for display: 8333 µs reads as `1/120`.
    var shutterDenominator: Int {
        guard exposureDurationUs > 0 else { return 0 }
        return Int((1_000_000.0 / Double(exposureDurationUs)).rounded())
    }
}

/// A partial mutation of `CameraState`. Only the keys the sender actually
/// touched are present, which is what makes concurrent phone/PC edits safe.
struct CameraMutation: Codable, Equatable, Sendable {
    var lensId: String?
    var zoom: Double?
    var lensLocked: Bool?

    var width: Int?
    var height: Int?
    var fps: Int?
    var codec: VideoCodec?
    var hdr: HDRMode?
    var stabilization: Stabilization?

    var exposureMode: ExposureMode?
    var iso: Double?
    var exposureDurationUs: Int?
    var ev: Double?
    var exposureLocked: Bool?

    var whiteBalanceMode: WhiteBalanceMode?
    var whiteBalancePreset: WhiteBalancePreset?
    var temperature: Double?
    var tint: Double?

    var focusMode: FocusMode?
    var focusPosition: Double?
    var focusLocked: Bool?
    var faceDrivenFocus: Bool?

    var torch: TorchMode?
    var torchLevel: Double?
    var mirrored: Bool?
    var orientation: OrientationLock?

    var brightness: Double?
    var contrast: Double?
    var saturation: Double?
    var warmth: Double?
    var sharpness: Double?

    var isEmpty: Bool {
        lensId == nil && zoom == nil && lensLocked == nil
            && width == nil && height == nil && fps == nil && codec == nil
            && hdr == nil && stabilization == nil
            && exposureMode == nil && iso == nil && exposureDurationUs == nil
            && ev == nil && exposureLocked == nil
            && whiteBalanceMode == nil && whiteBalancePreset == nil
            && temperature == nil && tint == nil
            && focusMode == nil && focusPosition == nil && focusLocked == nil
            && faceDrivenFocus == nil
            && torch == nil && torchLevel == nil && mirrored == nil && orientation == nil
            && brightness == nil && contrast == nil && saturation == nil
            && warmth == nil && sharpness == nil
    }

    /// Applies every present key onto `state`. Clamping to what the hardware
    /// supports happens afterwards, in `CameraStateStore`, because only it can
    /// see the live `AVCaptureDevice`.
    func apply(to state: inout CameraState) {
        if let v = lensId { state.lensId = v }
        if let v = zoom { state.zoom = v }
        if let v = lensLocked { state.lensLocked = v }

        if let v = width { state.width = v }
        if let v = height { state.height = v }
        if let v = fps { state.fps = v }
        if let v = codec { state.codec = v }
        if let v = hdr { state.hdr = v }
        if let v = stabilization { state.stabilization = v }

        if let v = exposureMode { state.exposureMode = v }
        if let v = iso { state.iso = v }
        if let v = exposureDurationUs { state.exposureDurationUs = v }
        if let v = ev { state.ev = v }
        if let v = exposureLocked { state.exposureLocked = v }

        if let v = whiteBalanceMode { state.whiteBalanceMode = v }
        if let v = whiteBalancePreset { state.whiteBalancePreset = v }
        if let v = temperature { state.temperature = v }
        if let v = tint { state.tint = v }

        if let v = focusMode { state.focusMode = v }
        if let v = focusPosition { state.focusPosition = v }
        if let v = focusLocked { state.focusLocked = v }
        if let v = faceDrivenFocus { state.faceDrivenFocus = v }

        if let v = torch { state.torch = v }
        if let v = torchLevel { state.torchLevel = v }
        if let v = mirrored { state.mirrored = v }
        if let v = orientation { state.orientation = v }

        if let v = brightness { state.brightness = v }
        if let v = contrast { state.contrast = v }
        if let v = saturation { state.saturation = v }
        if let v = warmth { state.warmth = v }
        if let v = sharpness { state.sharpness = v }
    }
}

/// What the PC receives. Deliberately unrelated to what the phone captures or
/// records — that separation is the whole point.
struct StreamProfile: Codable, Equatable, Sendable {
    var width: Int = 1280
    var height: Int = 720
    var fps: Int = 30
    var codec: VideoCodec = .h264
    var bitrate: Int = 6_000_000
    var keyframeIntervalSeconds: Double = 2

    static let webcam1080p30 = StreamProfile(width: 1920, height: 1080, fps: 30,
                                             codec: .h264, bitrate: 8_000_000)
    static let webcam720p30  = StreamProfile(width: 1280, height: 720, fps: 30,
                                             codec: .h264, bitrate: 4_000_000)
    static let webcam1080p60 = StreamProfile(width: 1920, height: 1080, fps: 60,
                                             codec: .h264, bitrate: 12_000_000)
}
